test/03_reduction/reduction_test.cpp

```diff
@@ -0,0 +1,419 @@
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+ * SOFTWARE.
+ *
+ *******************************************************************************/
+#include <hiptensor/hiptensor.hpp>
+
+#include "data_types.hpp"
+#include "logger.hpp"
+#include "reduction/reduction_cpu_reference.hpp"
+#include "reduction_test.hpp"
+#include "utils.hpp"
+#include "llvm/hiptensor_options.hpp"
+
+namespace hiptensor
+{
+    /*static*/ std::stringstream ReductionTest::sAPILogBuff = std::stringstream();
+
+    static void logMessage(int32_t logLevel, const char* funcName /*=""*/, const char* msg /*=""*/)
+    {
+        ReductionTest::sAPILogBuff << msg;
+    }
+
+    ReductionTest::ReductionTest()
+        : Base()
+    {
+        reset();
+        // Handle our own outputs
+        hiptensorLoggerOpenFile("/dev/null");
+        hiptensorLoggerSetCallback(logMessage);
+    }
+
+    // Kernel run checks. Virtual as different Reduction kernels have different requirements
+    // True = run test
+    // False = skip test
+    bool ReductionTest::checkDevice(hipDataType            datatype,
+                                    hiptensorComputeType_t computeDataType) const
+    {
+        return ((datatype == HIP_R_32F || computeDataType == HIP_R_32F) && isF32Supported())
+               || ((datatype == HIP_R_64F || computeDataType == HIP_R_64F) && isF64Supported());
+    }
+
+    bool ReductionTest::checkSizes() const
+    {
+        return true;
+    }
+
+    void ReductionTest::reset()
+    {
+        handle = nullptr;
+
+        mRepeats          = 1u;
+        mRunFlag          = true;
+        mValidationResult = false;
+        mMaxRelativeError = 0.0;
+    }
+
+    ReductionResource* ReductionTest::getResource() const
+    {
+        return DataStorage::instance().get();
+    }
+
+    void ReductionTest::SetUp()
+    {
+        // reset API log buffer
+        sAPILogBuff.str(std::string());
+
+        auto param      = Base::GetParam();
+        auto testType   = std::get<0>(param);
+        auto logLevel   = std::get<1>(param);
+        auto lengths    = std::get<2>(param);
+        auto outputDims = std::get<3>(param);
+        auto alpha      = std::get<4>(param);
+        auto beta       = std::get<5>(param);
+        auto op         = std::get<6>(param);
+
+        EXPECT_TRUE((lengths.size() > 1) && (lengths.size() <= 6));
+        EXPECT_TRUE((outputDims.size() >= 1) && (outputDims.size() <= 6));
+
+        EXPECT_TRUE((op == HIPTENSOR_OP_ADD) || (op == HIPTENSOR_OP_MUL) || (op == HIPTENSOR_OP_MAX)
+                    || (op == HIPTENSOR_OP_MIN));
+
+        EXPECT_EQ(testType.size(), 2); // HIP_R_16F or HIP_R_32F
+        auto acDataType      = testType[0];
+        auto computeDataType = convertToComputeType(testType[1]);
+        EXPECT_TRUE(
```

Comment:
Reduction instances only include f32 and f64 for now. Shall we disable the F16 and BF16 tests?

Reply:
Right now, CK has not fully supported bf16 and f16. There are building error to build these instances. So we cannot test f16 and bf16 now.